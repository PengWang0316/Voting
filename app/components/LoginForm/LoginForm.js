import React, { Component, Fragment } from 'react';
import { Text, View } from 'react-native';
import { connect } from 'react-redux';
import { Input, Icon, Button } from 'react-native-elements';

import { loginWithPassword, emptyUser } from '../../actions/UserActions';
// import LoginScreenContext from '../../contexts/LoginScreenContext';
import styles from './Styles';

type Props = {
  user: Object,
  loginWithPassword: Function,
  emptyUser: Function,
  handleSnackbarUpdate: Function,
  navigation: Object,
};

type States = {
  username: string,
  password: string,
  isReady: boolean,
  isError: boolean,
  loginWithPassword: Function,
  user: Object,
};

const LOGIN_FAIL_MESSAGE = 'Wrong username or password';
const COLOR_ERROR = '#ff1744';

/**
 * The login form component.
 */
export class LoginForm extends Component<Props, States> {
  state = {
    username: '',
    password: '',
    isReady: false,
    isSubmitted: false,
  };

  /**
   * Call the onToggleSnackbar when isSubmitted state is true and props received a null for user.
   * @param {object} nextProps contains the new props value.
   * @param {object} prevState contains the previous state's value.
   * @return {Object} Return an object to change the state.
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.isSubmitted && nextProps.user.isFail) {
      nextProps.handleSnackbarUpdate({
        isShowSnackbar: true, snackbarMessage: LOGIN_FAIL_MESSAGE, snackbarBgColor: COLOR_ERROR,
      });
      nextProps.emptyUser(); // Reset user state to an empty object. So, next time the back-end code still can send a isFail mark when the authentication fails.
      return { isSubmitted: false };
    }
    if (prevState.isSubmitted && nextProps.user._id) nextProps.navigation.goBack();
    return null;
  }

  /**
   * Call the backend to validate login information.
   * @return {null} No return.
   */
  handleLoginClick = () => this.setState(
    { isSubmitted: true },
    () => this.props.loginWithPassword({
      username: this.state.username,
      password: this.state.password,
    }),
  );

  /**
   * Update the username and isReady state when a user type in the text input.
   * @param {string} text is the string a user is typing.
   * @return {null} No return.
   */
  handleUsernameTextChange = text => this.setState(({ password }) => ({
    username: text,
    isReady: text.length !== 0 && password.length !== 0,
  }));

  /**
   * Update the password and isReady state when a user type in the text input.
   * @param {string} text is the string a user is typing.
   * @return {null} No return.
   */
  handlePasswordTextChange = text => this.setState(({ username }) => ({
    password: text,
    isReady: text.length !== 0 && username.length !== 0,
  }));

  /**
   * The render method for the component.
   * @return {jsx} Return jsx for the component.
   */
  render() {
    const {
      username, password, isReady, isSubmitted,
    } = this.state;

    return (
      <Fragment>

        <View>
          <Text style={styles.title}>Username Login</Text>
        </View>

        <View>
          <Input
            onChangeText={this.handleUsernameTextChange}
            placeholder="Username"
            value={username}
            leftIcon={(
              <Icon
                type="font-awesome"
                name="user-o"
                size={18}
                color="gray"
              />
            )}
          />
        </View>

        <View>
          <Input
            onChangeText={this.handlePasswordTextChange}
            value={password}
            secureTextEntry
            placeholder="Password"
            leftIcon={(
              <Icon
                type="feather"
                name="lock"
                size={18}
                color="gray"
              />
            )}
          />
        </View>

        <View style={styles.buttonView}>
          <Button
            disabled={isSubmitted || !isReady}
            title="LOGIN"
            onPress={this.handleLoginClick}
            buttonStyle={styles.loginBtn}
          />
        </View>

      </Fragment>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });

/* istanbul ignore next */
const mapDispatchToProprs = dispatch => ({
  loginWithPassword: params => dispatch(loginWithPassword(params)),
  emptyUser: () => dispatch(emptyUser()),
});
export default connect(mapStateToProps, mapDispatchToProprs)(LoginForm);
