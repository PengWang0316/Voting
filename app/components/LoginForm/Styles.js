import { StyleSheet } from 'react-native';

import Theme from '../../Theme';

export default StyleSheet.create({
  title: {
    fontWeight: 'bold',
    fontSize: 20,
    marginBottom: 15,
    color: Theme.primary.main,
  },
  buttonView: {
    marginTop: 30,
  },
  loginBtn: {
    backgroundColor: Theme.primary.main,
    borderRadius: 15,
  },
});
