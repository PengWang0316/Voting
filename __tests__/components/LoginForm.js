import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { LoginForm } from '../../app/components/LoginForm/LoginForm';
// import context from '../../app/contexts/LoginScreenContextTestHelper';

jest.mock('react-native-elements', () => ({
  Button: () => 'Button',
  Icon: () => 'Icon',
  Input: () => 'Input',
}));

describe('LoginForm', () => {
  const defaultProps = {
    user: {},
    handleSnackbarUpdate: jest.fn(),
    emptyUser: jest.fn(),
    loginWithPassword: jest.fn(),
    navigation: { goBack: jest.fn() },
  };
  const getShallowComponent = (props = defaultProps) => shallow(<LoginForm {...props} />);

  test('Initial states', () => {
    const component = getShallowComponent();
    const {
      username, password, isReady, isSubmitted,
    } = component.instance().state;

    expect(username).toBe('');
    expect(password).toBe('');
    expect(isReady).toBe(false);
    expect(isSubmitted).toBe(false);
  });

  test('getDerivedStateFromProps isSubmitted false', () => expect(LoginForm.getDerivedStateFromProps(null, { isSubmitted: false })).toBeNull());
  test('getDerivedStateFromProps isSubmitted is true and user.isFail is false', () => expect(LoginForm.getDerivedStateFromProps({ user: { isFail: false } }, { isSubmitted: true })).toBeNull());
  test('getDerivedStateFromProps isSubmitted is true and user.isFail is true', () => {
    const mockUpdateFn = jest.fn();
    const mockEmptyUserFn = jest.fn();
    const nextProps = {
      user: { isFail: true },
      handleSnackbarUpdate: mockUpdateFn,
      emptyUser: mockEmptyUserFn,
    };
    const prevState = { isSubmitted: true };
    expect(LoginForm.getDerivedStateFromProps(nextProps, prevState)).toEqual({ isSubmitted: false })
    expect(mockUpdateFn).toHaveBeenCalledTimes(1);
    expect(mockUpdateFn).toHaveBeenLastCalledWith({ isShowSnackbar: true, snackbarMessage: 'Wrong username or password', snackbarBgColor: '#ff1744' });
    expect(mockEmptyUserFn).toHaveBeenCalledTimes(1);
  });

  test('getDerivedStateFromProps isSubmitted is true and has user _id', () => {
    const mockGoBack = jest.fn();
    const nextProps = {
      user: { _id: true },
      navigation: { goBack: mockGoBack },
    };
    const prevState = { isSubmitted: true };
    expect(LoginForm.getDerivedStateFromProps(nextProps, prevState)).toEqual(null);
    expect(mockGoBack).toHaveBeenCalledTimes(1);
  });

  test('handleLoginClick', () => {
    const component = getShallowComponent();
    const someStates = { username: 'username', password: 'password' };
    component.setState(someStates);
    component.instance().handleLoginClick();
    expect(component.state('isSubmitted')).toBe(true);
    expect(defaultProps.loginWithPassword).toHaveBeenCalledTimes(1);
    expect(defaultProps.loginWithPassword).toHaveBeenLastCalledWith(someStates);
  });

  test('handleUsernameTextChange text length equal 0', () => {
    const component = getShallowComponent();
    const text = '';
    component.instance().handleUsernameTextChange(text);
    expect(component.state('username')).toBe(text);
    expect(component.state('isReady')).toBe(false);
  });

  test('handleUsernameTextChange passowrd length equal 0', () => {
    const component = getShallowComponent();
    const text = 'text';
    component.setState({ password: '' });
    component.instance().handleUsernameTextChange(text);
    expect(component.state('username')).toBe(text);
    expect(component.state('isReady')).toBe(false);
  });

  test('handleUsernameTextChange username and password length not equal 0', () => {
    const component = getShallowComponent();
    const text = 'text';
    component.setState({ password: 'password' });
    component.instance().handleUsernameTextChange(text);
    expect(component.state('username')).toBe(text);
    expect(component.state('isReady')).toBe(true);
  });

  test('handlePasswordTextChange password length equal 0', () => {
    const component = getShallowComponent();
    const password = '';
    component.instance().handlePasswordTextChange(password);
    expect(component.state('password')).toBe(password);
    expect(component.state('isReady')).toBe(false);
  });

  test('handlePasswordTextChange username length equal 0', () => {
    const component = getShallowComponent();
    const password = 'password';
    component.setState({ username: '' });
    component.instance().handlePasswordTextChange(password);
    expect(component.state('password')).toBe(password);
    expect(component.state('isReady')).toBe(false);
  });

  test('handlePasswordTextChange username and password length not equal 0', () => {
    const component = getShallowComponent();
    const password = 'password';
    component.setState({ username: 'username' });
    component.instance().handlePasswordTextChange(password);
    expect(component.state('password')).toBe(password);
    expect(component.state('isReady')).toBe(true);
  });

  test('Snapshot', () => expect(renderer.create(<LoginForm {...defaultProps} />).toJSON()).toMatchSnapshot());
});
