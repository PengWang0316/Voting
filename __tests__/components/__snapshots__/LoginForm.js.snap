// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Snapshot 1`] = `
Array [
  <View>
    <Text
      style={
        Object {
          "color": "#81c784",
          "fontSize": 20,
          "fontWeight": "bold",
          "marginBottom": 15,
        }
      }
    >
      Username Login
    </Text>
  </View>,
  <View>
    Input
  </View>,
  <View>
    Input
  </View>,
  <View
    style={
      Object {
        "marginTop": 30,
      }
    }
  >
    Button
  </View>,
]
`;
